OpenRA.Mods.Common/Traits/BotModules/PowerDownBotMoudle.cs

```diff
@@ -0,0 +1,148 @@
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages AI powerdown.")]
+	public class PowerDownBotMoudleInfo : ConditionalTraitInfo
+	{
+		[Desc("Delay (in ticks) between toggle")]
+		public readonly int Interval = 150;
+
+		[Desc("ConditionToken that powerdown")]
+		public readonly string PowerdownCondition = "powerdown";
+
+		public override object Create(ActorInitializer init) { return new PowerDownBotMoudle(init.Self, this); }
+	}
+
+	public class PowerDownBotMoudle : ConditionalTrait<PowerDownBotMoudleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		PowerManager playerPower;
+		int toggleTick;
+
+		public PowerDownBotMoudle(Actor self, PowerDownBotMoudleInfo info)
+			: base(info)
+		{
+			world = self.World;
+
+			if (world.Type == WorldType.Editor)
+				return;
+
+			player = self.Owner;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			toggleTick = world.LocalRandom.Next(0, Info.Interval);
+			if (player == null || player.PlayerActor == null)
+			{
+				playerPower = null;
+				return;
+			}
+
+			playerPower = player.PlayerActor.TraitOrDefault<PowerManager>();
+		}
+
+		protected SortedList<int, List<Actor>> GetToggleableBuildings(IBot bot)
+		{
+			SortedList<int, List<Actor>> toggleableBuildings = new SortedList<int, List<Actor>>();
+
+			var toggleable = bot.Player.World.ActorsHavingTrait<ToggleConditionOnOrder>(t => !t.IsTraitDisabled && !t.IsTraitPaused && t.Info.Condition == Info.PowerdownCondition)
+				.Where(a => a != null && !a.IsDead && a.Owner == bot.Player && a.Info.HasTraitInfo<PowerInfo>() && a.Info.HasTraitInfo<PowerMultiplierInfo>() && a.Info.HasTraitInfo<BuildingInfo>());
+
+			foreach (var a in toggleable)
+			{
+				var powerTarit = a.TraitsImplementing<Power>().Where(t => !t.IsTraitDisabled).ToArray();
+				var powerMulTrait = a.TraitsImplementing<PowerMultiplier>().ToArray();
+				if (powerTarit.Any())
+				{
+					var powerChanging = a.TraitsImplementing<Power>().Sum(p => p.Info.Amount) * (a.TraitsImplementing<PowerMultiplier>().Sum(p => p.Info.Modifier) - 100) / 100;
+
+					if (powerMulTrait.Where(t => !t.IsTraitDisabled).Any())
+						powerChanging = -powerChanging;
+					List<Actor> powerList;
+					var hasPowerValue = toggleableBuildings.TryGetValue(powerChanging, out powerList);
+					if (hasPowerValue)
+						powerList.Add(a);
+					else
+					{
+						powerList = new List<Actor>();
+						powerList.Add(a);
+						toggleableBuildings.Add(powerChanging, powerList);
+					}
+				}
+			}
+
+			return toggleableBuildings;
+		}
+
+		public void BotTick(IBot bot)
+		{
+			if (toggleTick > 0 || playerPower == null)
+			{
+				toggleTick--;
+				return;
+			}
+
+			var toggleableBuildings = GetToggleableBuildings(bot);
```

Comment:
This looks needlessly complicated. Is there a reason for not simply (a) returning an `IEnumerable<TraitPair<Power>>` here, sorted by largest-least power drain, and (b) keeping a separate `List<TraitPair<Power>>` on the object to track what has been powered down to restore later?

 of powerdown-able

Reply:
Because it may be fast,  I guess.  By using `SortedList`,  I can skip those who cause same power change when searching through (it is possible of AI build a lot of same type of defences) meanwhile `SortedList` can sort those classified group in order.

Reply:
This should also be only generated when we actually need to power anything down. Keeping a list of what has been powered down will save that lookup as well and thus be faster.

Reply:
> track what has been powered down to restore later? 
As for this, the owner can change and the actor can die, if AI captures a powerdown building this track will not affect this new building. I think rescan can be better at this. I get a better idea on optimization, though. I can scan due to the need of upper calling.

Reply:
1) Why would you need to skip over objects when powering down?

2) Capturing isn't enabled in most mods, and in any case we should really be restoring power to a building if it changes owner. Fixing this would make the code here much simpler.

Reply:
> Why would you need to skip over objects when powering down?

For example, we have N x -40 defences (may be different) needs to be turn on but we only have 70 extra power for that. By using my idea here we turn on only one of the -40 defences, skip the rest and search -20 or -10 groups that has already in list.

Reply:
That adds a significant complexity overhead for I expect little observable benefit ingame. Can we please use a simpler approach to keep the code understandable and maintainable for the future?

Reply:
Another problem. On trace, how can we find out the units that are dead/ owner changed meanwhile delete them safely? I used to make a trace in fact but it give me "Collection was modified; enumeration operation may not execute" in `foreach`. I know I have to use `for` for this but I think it will also be complex to look at. 

In this case I believe rescan is better. and thus we don't need to fix toggle on owner change.

Reply:
Take a look at `CaptureManagerBotModule` and `HarvesterManagerBotModule` to see how to clear stale actor references.

Reply:
fixed.  but captured toggled building will now not listen to the order.